sway-lib-std/src/ecr.sw

```diff
@@ -1,31 +1,44 @@
 library ecr;
 
-use ::b512::B512;
 use ::address::Address;
+use ::b512::B512;
+use ::constants::ZERO;
+use ::hash::{HashMethod, hash_pair};
+use ::result::Result;
+
+pub enum EcRecoverError {
+    UnrecoverableAddress: (),
+    UnrecoverablePublicKey: (),
+}
 
 /// Recover the public key derived from the private key used to sign a message
-pub fn ec_recover(signature: B512, msg_hash: b256) -> B512 {
+pub fn ec_recover(signature: B512, msg_hash: b256) -> Result<B512, EcRecoverError> {
     let public_key = ~B512::new();
 
     asm(buffer: public_key.bytes, sig: signature.bytes, hash: msg_hash) {
         ecr buffer sig hash;
     };
-
-    public_key
+    /// if recovered key is empty
+    if (public_key.bytes)[0] == ZERO && (public_key.bytes)[1] == ZERO {
+        Result::Err(EcRecoverError::UnrecoverablePublicKey)
+    } else {
+        Result::Ok(public_key)
+    }
 }
 
 /// Recover the address derived from the private key used to sign a message
-pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Address {
-    let address = asm(sig: signature.bytes, hash: msg_hash, addr_buffer, pub_key_buffer, hash_len: 64) {
-        move addr_buffer sp; // Buffer for address.
-        cfei i32;
-        move pub_key_buffer sp; // Temporary buffer for recovered key.
-        cfei i64;
-        ecr pub_key_buffer sig hash; // Recover public_key from sig & hash.
-        s256 addr_buffer pub_key_buffer hash_len; // Hash 64 bytes to the addr_buffer.
-        cfsi i64; // Free temporary key buffer.
-        addr_buffer: b256
-    };
+pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Result<Address, EcRecoverError> {
+    let pub_key_result = ec_recover(signature, msg_hash);
 
-    ~Address::from(address)
+    if let Result::Ok(p) = pub_key_result {
+        let address = hash_pair((p.bytes)[0], (p.bytes)[1], HashMethod::Sha256);
+        if address != ZERO {
+            Result::Ok(~Address::from(address))
+        } else {
+            Result::Err(EcRecoverError::UnrecoverableAddress)
+        }
+    } else {
+        // propagate the error if it exists
+        Result::Err(EcRecoverError::UnrecoverablePublicKey)
```

Comment:
```suggestion
        pub_key_result
```
Maybe one day `ec_recover()` might return a different error?  Might as well actually propagate the original result.

Reply:
FWIW, I was able to propagate the original error with:
`Result::Err(e)`
but not by just returning
`pub_key_result`, even though they are of the same type.

Reply:
Argh, really?  What was the error?

Reply:
```
✦ ❯ cargo run -- ec_recover_test
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
     Running `/Users/furnic/dev/fuel/sway/target/debug/test ec_recover_test`
 Compiling should_pass/stdlib/ec_recover_test
  Compiled library "core".
error
  --> sway-lib-std/src/result.sw:16:5
   |
14 |
15 |     /// Contains the success value
16 |     Ok: T,
   |     ^^ Mismatched types.
expected: struct Address<b256>
found:    struct B512<[b256; 2]>.
help: `then` branch must return expected type.
17 |
18 |     /// Contains the error value
   |
____

error
  --> sway-lib-std/src/result.sw:16:5
   |
14 |
15 |     /// Contains the success value
16 |     Ok: T,
   |     ^^ Mismatched types.
expected: struct B512<[b256; 2]>
found:    struct Address<b256>.
help: `else` branch must return expected type.
17 |
18 |     /// Contains the error value
   |
____

error
  --> sway-lib-std/src/result.sw:16:5
   |
14 |
15 |     /// Contains the success value
16 |     Ok: T,
   |     ^^ Mismatched types.
expected: struct Address<b256>
found:    struct B512<[b256; 2]>.
help: Implicit return must match up with block's type.
17 |
18 |     /// Contains the error value
   |
____

error
  --> sway-lib-std/src/result.sw:16:5
   |
14 |
15 |     /// Contains the success value
16 |     Ok: T,
   |     ^^ Mismatched types.
expected: struct Address<b256>
found:    struct B512<[b256; 2]>.
help: Function body's return type does not match up with its return type annotation.
17 |
18 |     /// Contains the error value
   |
____

  Aborting due to 4 errors.
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Failed to compile std', test/src/e2e_vm_tests/harness.rs:69:46
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

Reply:
This is weird because they're literally the same type...

Reply:
Ahh, no, sorry, this makes sense.  I should've looked closer -- the two functions return different types, so it should error.  We could maybe make a better error for it one day.

When we have `Result::map()` (I assume we don't yet) then this could be simplified into something like:
```rust
pub fn ec_recover_address(signature: B512, msg_hash: b256) -> Result<Address, EcRecoverError> {
    ec_recover(signature, msg_hash).map(|pub_key| {
        let address = hash_pair((pub_key.bytes)[0], (pub_key.bytes)[1], HashMethod::Sha256);
        ~Address::from(address)
    })
}
```
This converts the type from `B512` to `Address` but keeps the error the same.

Reply:
I don't think we even have closures yet, so this is a long way off... :)  Sorry for wasting your time on this.

Reply:
No worries. Yeah `Result::map()` looks nice, I've never used it in Rust, but good to know it's a thing.